File: engine/building.lua
local Building = {}
Building.__index = Building

function Building:new(layout, x, y)
   local self = setmetatable({}, Building)
   self.x, self.y = x, y
   self.layout = layout
   self.w = #self.layout[1][1]
   self.h = #self.layout[1]
   return self
end

function Building:tiles(x, y, basic_block)
   local i = 1
   local stack = {}
   while self.layout[i] do
      local block = self.layout[self.layout[i][y+1]:sub(x+1,x+1)]
      if block == Block.NOTHING and i == 1 and basic_block then block = basic_block end
      stack[#stack+1] = block
      i = i+1
   end
   return stack
end

-------------
-- PRIVATE --
-------------

function Building:__tostring()
   return '[Building]'
end

return Building

-- vim: ts=3:sw=3:sts=3:expandtab

File: engine/building.cc
#include "engine/building.h"

Building::Building(int x, int y, BuildingType type, int n)
    : X(x), Y(y), type(type), 
      layout(BuildingLayouts[BuildingPair(type, n)])
{
}


int 
Building::Tiles(Block (&block)[10], int x, int y) const
{
    for(int i=0; i<layout.floors; i++) {
        unsigned int ui = static_cast<unsigned int>(i),
                     uy = static_cast<unsigned int>(y),
                     ux = static_cast<unsigned int>(x);
        const char blk = layout.floor[ui][uy][ux];
        block[i] = layout.label.at(blk);
    }
    return layout.floors;
}


// vim: ts=4:sw=4:sts=4:expandtab
